src/tunnel.cc

```diff
@@ -1274,17 +1269,9 @@ TunnelStateData::noteDestinationsEnd(ErrorState *selectionError)
     PeerSelectionInitiator::subscribed = false;
     destinations->destinationsFinalized = true;
     if (!destinationsFound) {
-
-        // XXX: Honor clientExpectsConnectResponse() before replying.
-
-        if (selectionError)
-            return sendError(selectionError, "path selection has failed");
-
-        if (savedError)
-            return sendError(savedError, "all found paths have failed");
-
-        return sendError(new ErrorState(ERR_CANNOT_FORWARD, Http::scInternalServerError, request.getRaw(), al),
-                         "path selection found no paths");
+        assert(!savedError);
+        return sendErrorAndDestroy(selectionError, selectionError ?
```

Comment:
This is the only place where the error argument may be nil. In this case, I think, we should create the error here (moving that code from sendErrorAndDestroy()).
We can probably simplify sendErrorAndDestroy() itself, eliminating the error argument and always calling saveError() before sendErrorAndDestroy() (except for cases when savedError passed as an argument). In this case, sendErrorAndDestroy() would deal with only savedError and finalError and become simpler.

Reply:
> We can probably simplify sendErrorAndDestroy() itself eliminating the error argument

Done at a0d34b7. If we find problems with this (arguable) change, I'll revert it.